functions/src/index.ts

```diff
@@ -0,0 +1,161 @@
+    if (readState == "RECEIVED" || readState == "READ") {
+      return;
+    }
+
+    const [recipientTokenSnapshot,
+      firstNameSenderSnapshot,
+      lastNameSenderSnapshot]: DataSnapshot[] =
+        await fetchSnapshotValues(change, senderWithCommas, recipient);
+
+    // if no token found for given recipient (i.e., .val() returns null)
+    // return without sending push notification
+    if (!recipientTokenSnapshot.val()) {
+      return;
+    }
+
+    const payload = createPayload(
+      firstNameSenderSnapshot.val(), lastNameSenderSnapshot.val(),
+      message.content, sender);
+
+    // send push notification
+    await admin.messaging().sendToDevice(recipientTokenSnapshot.val(), payload);
+
+    // update readState to RECEIVED
+    await change.after.ref.update({readState: "RECEIVED"});
+  });
+
+
+/**
+ * Helper function that fetches snapshot values for the recipient token,
+ * sender's first name, and sender's last name from the database in parallel.
+ *
+ * @param {Change<DataSnapshot>} change The database change object.
+ * @param {string} senderWithCommas The sender ID with dots replaced by commas.
+ * @param {string} recipientWithCommas The ID of the recipient with
+ *   dots replaced by commas.
+ * @return {Promise<DataSnapshot[]>} An object containing the
+ *   recipient token snapshot,
+ * sender's first name snapshot, and sender's last name snapshot.
+ */
+async function fetchSnapshotValues(
+  change: Change<DataSnapshot>,
+  senderWithCommas: string,
+  recipientWithCommas: string,
+): Promise<DataSnapshot[]> {
+  const [recipientTokenSnapshot,
+    firstNameSenderSnapshot,
+    lastNameSenderSnapshot] =
+      await Promise.all([
+        change.after.ref.root
+          .child("/coachme/fcmTokens/" + recipientWithCommas)
+          .once("value"),
+        change.after.ref.root
+          .child("/coachme/accounts/" + senderWithCommas + "/firstName")
+          .once("value"),
+        change.after.ref.root
+          .child("/coachme/accounts/" + senderWithCommas + "/lastName")
+          .once("value"),
+      ]);
+
+  return [recipientTokenSnapshot,
+    firstNameSenderSnapshot,
+    lastNameSenderSnapshot];
+}
+
+
+/**
+ * Helper function that creates the payload for the push notification.
+ *
+ * @param {string} firstNameSender - The first name of the sender.
+ * @param {string} lastNameSender - The last name of the sender.
+ * @param {string} bodyContent - The message content.
+ * @param {string} sender - The ID of the sender.
+ * @return {admin.messaging.MessagingPayload} - The payload object.
+ */
+function createPayload(
+  firstNameSender: string,
+  lastNameSender: string,
+  bodyContent: string,
+  sender: string): admin.messaging.MessagingPayload {
+  return {
+    notification: {
+      title: `${firstNameSender} ${lastNameSender}`,
+      body: bodyContent,
+      // needed to tell the app to open the chat activity
+      click_action: "OPEN_CHAT_ACTIVITY",
```

Comment:
You add this here but you don't seem to use it in onMessageReceived of InAppNotificationService. Is it still necessary ? In sendMessagingNotification you seem to re-add it manually (line 64 of InAppNotifier.kt)

Reply:
This is needed for the case where the push notification is received when the user is not inside the application. I.e., the InAppNotificationService only runs when the app is in the foreground. Otherwise the act of clicking on the push notification just takes the user to the login activity with the given action passed in the intent. Therefore, in order to be able to decide whether or not the user just opened the application or pressed on a messaging push notification, this click action is needed and checked inside the login activity.

I have adapted the description of the InAppNotificationService accordingly